src/components/Todo/Todo.tsx

```diff
@@ -0,0 +1,258 @@
+import React, { Dispatch, SetStateAction, useState, useRef } from 'react';
+import { TodoInterface } from '../../types/Todo';
+import { deleteTodo, updateTodo } from '../../api/todos';
+
+interface Props {
+  todos: TodoInterface[];
+  refreshTodos: Dispatch<SetStateAction<TodoInterface[]>>;
+  setErrorMsg: (arg: string) => void;
+}
+
+const TodoComponent: React.FC<Props> = ({
+  todos: initialTodos,
+  refreshTodos,
+  setErrorMsg,
+}) => {
+  const [todoClickToEdit, setTodoClickToEdit] = useState(-1);
+  const [valueInInputForEditTodo, setValueInInputForEditTodo] = useState('');
+  const [requestRunning, setRequestRunning] = useState(true);
+  const refTodoToEdit = useRef<HTMLInputElement>(null);
+
+  const handleOnclickCompletTodo = (idx: number) => {
+    return () => {
+      refreshTodos(todos => {
+        const newTodos = [...todos];
+
+        newTodos[idx].isAwaitServer = true;
+
+        return newTodos;
+      });
+      updateTodo(initialTodos[idx].id, {
+        completed: !initialTodos[idx].completed,
+      })
+        .then(res => {
+          refreshTodos(todos => {
+            const newTodos = [...todos];
+
+            newTodos[idx] = res as TodoInterface;
+            newTodos[idx].isAwaitServer = false;
+
+            return newTodos;
+          });
+        })
+        .catch(() => {
+          setErrorMsg('Unable to update a todo');
+          refreshTodos(todos => {
+            const newTodos = [...todos];
+
+            newTodos[idx].isAwaitServer = false;
+
+            return newTodos;
+          });
+        });
+    };
+  };
+
+  const handleOnDeleteButtonTodo = (idx: number) => {
+    return () => {
+      refreshTodos(todos => {
+        const newTodos = [...todos];
+
+        newTodos[idx].isAwaitServer = true;
+
+        return newTodos;
+      });
+
+      deleteTodo(initialTodos[idx].id)
+        .then(() => {
+          refreshTodos(todos => {
+            const newTodos = todos.filter((_, i) => i !== idx);
+
+            return newTodos;
+          });
+        })
+        .catch(() => {
+          setErrorMsg('Unable to delete a todo');
+        });
+    };
+  };
+
+  const saveData = (idx: number) => {
+    return () => {
+      let indexTodoToBeUpdated: number;
+
+      refreshTodos(prev => {
+        indexTodoToBeUpdated = prev.findIndex(
+          e => e.id === initialTodos[idx].id,
+        );
+        const newPrev = [...prev];
+
+        newPrev[indexTodoToBeUpdated].isAwaitServer = true;
+
+        return newPrev;
+      });
+
+      if (valueInInputForEditTodo === initialTodos[idx].title) {
+        setTodoClickToEdit(-1);
+        setValueInInputForEditTodo('');
+
+        return;
+      }
+
+      if (valueInInputForEditTodo !== '') {
+        setRequestRunning(false);
```

Comment:
Why is requestRunning false when the request starts, and true when it finishes?
Either it is switched or the naming for this state is really bad